.kokoro/nightly/downstream-protobuf-binary-compatibility.sh

```diff
@@ -0,0 +1,107 @@
+#!/bin/bash
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eo pipefail
+
+# Comma-delimited list of repos to test with the local java-shared-dependencies
+if [ -z "${REPOS_UNDER_TEST}" ]; then
+  echo "REPOS_UNDER_TEST must be set to run downstream-protobuf-binary-compatibility.sh"
+  echo "Expects a comma-delimited list: i.e REPOS_UNDER_TEST=\"java-bigtable,java-bigquery\""
+  exit 1
+fi
+
+# Version of Protobuf-Java runtime to compile with
+if [ -z "${PROTOBUF_RUNTIME_VERSION}" ]; then
+  echo "PROTOBUF_RUNTIME_VERSION must be set to run downstream-protobuf-binary-compatibility.sh"
+  echo "Expects a single Protobuf-Java runtime version i.e. PROTOBUF_RUNTIME_VERSION=\"4.28.3\""
+  exit 1
+fi
+
+# Create two mappings of possible API names (Key: Maven Artifact ID Prefix, Value: Maven Group ID)
+# for the libraries that should be tested.
+# 1. These are special handwritten libraries in google-cloud-java that should be tested
+declare -A monorepo_handwritten_libraries
+monorepo_handwritten_libraries["grafeas"]="io.grafeas"
+monorepo_handwritten_libraries["google-cloud-vertexai"]="com.google.cloud"
+monorepo_handwritten_libraries["google-cloud-resourcemanager"]="com.google.cloud"
+
+# 2. These are the mappings of all the downstream handwritten libraries' artifacts
+declare -A downstream_handwritten_libraries
+downstream_handwritten_libraries["google-cloud"]="com.google.cloud"
+
+# Builds a string output to `artifact_list`. It contains a comma separate list of Maven GAV coordinates. Parses
+# the `versions.txt` file by searching for the matching artifact_id_prefix to get the corresponding version.
+function build_artifact_list() {
```

Comment:
For future enhancements: we could extract this function to a separate workflow, so that the artifact list can be passed to the binary check in a matrix, which can then be run in parallel. Similar to how we generate [Apiary libraries](https://github.com/googleapis/google-api-java-client-services/blob/main/.github/workflows/codegen.yaml#L9-L25).

Reply:
This function only builds a string of packages for the linkage checker to test again. The artifactList corresponds to modules found in a repo and every repo is already built using a matrix. 

ie.. bigtable -> artifactList: google-cloud-bigtable
google-cloud-java -> artifactList: translate,resourcemanager,vertexai

Reply:
Maybe I'm misunderstanding something, I thought there is still a [for loop](https://github.com/googleapis/sdk-platform-java/pull/3650/files#diff-561e09c5712d2899ae73c0d15235cfa6b1616cb571b3010e91b465c8d6c144a4R78) that runs each repo one by one?
Maybe I was not clear, when I mentioned "we could extract this function to a separate workflow", I meant that "we could extract the functionality of building a list of artifact for all repos to a separate workflow". 

i.e.  
```
google-cloud-java, java-bigtable, java-bigquery, java-bigquerystorage etc.
```
 -> 
```
google-cloud-bigtable, translate,resourcemanager,vertexai etc.
```

Then we put all of the artifacts to a Github matrix so that they can run in parallel.

Reply:
This job is already running in parallel and each downstream repo has it's own job in the matrix. The for loop is only building a list of source artifacts for the linkage checker to use since each repo can have multiple relevant artifacts to test

Reply:
Discussed offline that we are going to maintain a hardcode list of `repo -> list of artifacts` mapping. This makes the shell script much easier to maintain.